libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+#include <utility>
+#include <variant>
+#include <vector>
+
+namespace mamba::util
+{
+    /**
+     * In array binary tree.
+     *
+     * A binary tree, each node is either a leaf, or a node with exactly two children.
+     * This data structure is light and nothing prevents the user from representing
+     * any kind of binary directed acyclic graph (e.g. there can be multiple trees,
+     * or nodes could have multiple parents).
+     */
+    template <typename Branch, typename Leaf>
+    class flat_binary_tree
+    {
+    public:
+
+        using branch_type = Branch;
+        using leaf_type = Leaf;
+        struct BranchNode
+        {
+            branch_type data;
+            std::size_t left_child = 0;
+            std::size_t right_child = 0;
+        };
+        using branch_node = BranchNode;
+        using leaf_node = leaf_type;
+        using node_type = std::variant<branch_node, leaf_node>;
+        using node_list = std::vector<node_type>;
+        using size_type = typename node_list::size_type;
+        using idx_type = size_type;
+
+        [[nodiscard]] auto size() const -> size_type;
+        [[nodiscard]] auto empty() const -> bool;
+
+        void clear();
+        void reserve(size_type size);
+
+        auto add_leaf(const leaf_type& leaf) -> idx_type;
+        auto add_leaf(leaf_type&& leaf) -> idx_type;
+        auto add_branch(const branch_type& branch, idx_type left_child, idx_type right_child)
+            -> idx_type;
+        auto add_branch(branch_type&& branch, idx_type left_child, idx_type right_child) -> idx_type;
+
+        [[nodiscard]] auto node(idx_type idx) const -> const node_type&;
+        [[nodiscard]] auto node(idx_type idx) -> node_type&;
+        [[nodiscard]] auto is_branch(idx_type idx) const -> bool;
+        [[nodiscard]] auto is_leaf(idx_type idx) const -> bool;
+        [[nodiscard]] auto leaf(idx_type idx) const -> const leaf_type&;
+        [[nodiscard]] auto leaf(idx_type idx) -> leaf_type&;
+        [[nodiscard]] auto branch(idx_type idx) const -> const branch_type&;
+        [[nodiscard]] auto branch(idx_type idx) -> branch_type&;
+        [[nodiscard]] auto left(idx_type idx) const -> idx_type;
+        [[nodiscard]] auto right(idx_type idx) const -> idx_type;
+        [[nodiscard]] auto root() const -> idx_type;
+
+    private:
+
+        node_list m_nodes;
+        idx_type m_root = 0;
+
+        template <typename L>
+        auto add_leaf_impl(L&& leaf) -> idx_type;
+        template <typename B>
+        auto add_branch_impl(B&& branch, idx_type left_child, idx_type right_child) -> idx_type;
+    };
+
+    /**
+     * A parser for postfix expressions.
+     *
+     * The parser creates an expression tree and validate that the expression being pushed
+     * is a valid postfix expression.
+     * For example, for the expression ``a + b * c`` on might push ``a b c * +``
+     * or ``b c * a +``.
+     */
+    template <typename Variable, typename Operator>
+    class PostfixParser
```

Comment:
So this header is not "just" for flat_binary_tree, is it nammed aptly for everything it contains?

Reply:
In a sense, the goal of a Parser is a fat/stateful constructor for the tree.
The parser only goal is to create a tree, and tree are difficult to create without a parser.

Not sure what name would be best.